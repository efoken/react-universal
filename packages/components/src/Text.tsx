'use client';

import { OverridableComponent, SxProps, styled } from '@universal-ui/styles';
import { forwardRef } from 'react';
import {
  Platform,
  Text as RNText,
  TextProps as RNTextProps,
} from 'react-native';

export interface TextProps extends RNTextProps {
  sx?: SxProps;
}

const TextRoot = styled(RNText)();

export const Text = forwardRef<RNText, TextProps>(
  (
    {
      accessibilityElementsHidden,
      'aria-hidden': ariaHidden,
      importantForAccessibility,
      ...props
    },
    ref,
  ) => (
    <TextRoot
      ref={ref}
      {...(Platform.OS === 'web'
        ? { 'aria-hidden': ariaHidden }
        : {
            accessibilityElementsHidden:
              ariaHidden ?? accessibilityElementsHidden,
            importantForAccessibility: ariaHidden
              ? 'no-hide-descendants'
              : importantForAccessibility,
          })}
      {...props}
    />
  ),
) as OverridableComponent<TextProps, typeof RNText>;
