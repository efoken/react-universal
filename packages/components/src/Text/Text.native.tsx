import { normalizeLayoutEvent, normalizeRole, styled } from '@universal-ui/core';
import { forwardRef } from 'react';
import { Text as RNText } from 'react-native';
import type { TextProps, TextType } from './Text.types';

const TextRoot = styled(RNText, {
  name: 'Text',
  slot: 'Root',
})(({ theme }) => ({
  fontFamily: theme.fonts.body.family,
  fontStyle: 'normal',
  fontWeight: 'normal',
  position: 'static',
}));

export const Text = forwardRef<any, TextProps>(
  ({ 'aria-hidden': ariaHidden, lang, onLayout, role, style, ...props }, ref) => (
    <TextRoot
      ref={ref}
      accessibilityElementsHidden={ariaHidden}
      accessibilityLanguage={lang}
      aria-hidden={ariaHidden}
      importantForAccessibility={ariaHidden ? 'no-hide-descendants' : undefined}
      role={normalizeRole(role)}
      style={style as any}
      onLayout={normalizeLayoutEvent(onLayout)}
      {...props}
    />
  ),
) as unknown as TextType;

Text.displayName = 'Text';
