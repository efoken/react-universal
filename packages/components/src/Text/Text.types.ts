import type {
  AccessibilityProps,
  LayoutEvent,
  PlatformMethods,
  RNStyle,
  StyleProp,
  SxProps,
} from '@react-universal/core';
import type {
  AccessibilityProps as RNAccessibilityProps,
  TextProps as RNTextProps,
} from 'react-native';

export interface TextMethods extends PlatformMethods {}

export type TextStyle = Omit<RNStyle, 'objectFit'>;

export interface TextProps
  extends Omit<
      RNTextProps,
      keyof RNAccessibilityProps | 'nativeID' | 'onLayout' | 'selectable' | 'style'
    >,
    AccessibilityProps {
  as?: React.ElementType;
  dataSet?: Record<string, any>;
  dir?: 'ltr' | 'rtl' | 'auto';
  href?: string;
  hrefAttrs?: {
    download?: any;
    rel?: string;
    target?: React.HTMLAttributeAnchorTarget;
  };
  lang?: Intl.UnicodeBCP47LocaleIdentifier;
  onClick?: NonNullable<RNTextProps['onPress']>;
  /**
   * Callback invoked on mount and layout changes.
   */
  onLayout?: (event: LayoutEvent) => void;
  style?: StyleProp<TextStyle>;
  /**
   * The system prop that allows defining system overrides as well as additional
   * CSS styles.
   */
  sx?: SxProps;
}

export type TextType = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<TextProps> & React.RefAttributes<HTMLElement & TextMethods>
>;

export type TextOwnerState = Pick<TextProps, 'numberOfLines' | 'role'> & {
  hasTextAncestor: boolean;
  pressable: boolean;
};
