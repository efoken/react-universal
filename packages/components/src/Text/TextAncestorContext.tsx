'use client';

import { createContext } from 'react';

export const TextAncestorContext = createContext(false);
