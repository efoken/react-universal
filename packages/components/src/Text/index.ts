export { Text } from './Text';
export type { TextMethods, TextProps, TextStyle, TextType } from './Text.types';
