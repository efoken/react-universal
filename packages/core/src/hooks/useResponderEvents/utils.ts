const keyName = '__reactResponderId';

function isSelectionValid() {
  const selection = window.getSelection()!;
  const string = selection.toString();
  return (
    string.length > 0 &&
    string !== '\n' &&
    (selection.anchorNode?.nodeType === window.Node.TEXT_NODE ||
      selection.focusNode?.nodeType === window.Node.TEXT_NODE)
  );
}

function getEventPath(domEvent: Event): any[] {
  // The `selectionchange` event always has the document as the target. Use the
  // anchor node as the initial target to reconstruct a path. (We actually only
  // need the first responder node in practice.)
  if (domEvent.type === 'selectionchange') {
    const target = window.getSelection()?.anchorNode;
    return composedPathFallback(target);
  }
  return domEvent.composedPath != null
    ? domEvent.composedPath()
    : composedPathFallback(domEvent.target);
}

function composedPathFallback(target: EventTarget | Node | null | undefined): any[] {
  const path = [];
  while (target != null && target !== document.body) {
    path.push(target);
    target = (target as any).parentNode;
  }
  return path;
}

/**
 * Retrieve the responder ID from a host node.
 */
function getResponderId(node: any): number | undefined {
  if (node != null) {
    return node[keyName];
  }
}

/**
 * Store the responder ID on a host node.
 */
export function setResponderId(node: any, id: number) {
  if (node != null) {
    node[keyName] = id;
  }
}

/**
 * Filter the event path to contain only the nodes attached to the responder
 * system.
 */
export function getResponderPaths(domEvent: any): { idPath: number[]; nodePath: any[] } {
  const idPath: number[] = [];
  const nodePath = [];
  const eventPath = getEventPath(domEvent);
  for (const node of eventPath) {
    const id = getResponderId(node);
    if (id != null) {
      idPath.push(id);
      nodePath.push(node);
    }
  }
  return { idPath, nodePath };
}

/**
 * Walk the paths and find the first common ancestor.
 */
export function getLowestCommonAncestor(pathA: any[], pathB: any[]): any {
  let pathALength = pathA.length;
  let pathBLength = pathB.length;
  if (
    // If either path is empty
    pathALength === 0 ||
    pathBLength === 0 ||
    // If the last elements aren't the same there can't be a common ancestor
    // that is connected to the responder system
    pathA[pathALength - 1] !== pathB[pathBLength - 1]
  ) {
    return null;
  }

  let itemA = pathA[0];
  let indexA = 0;
  let itemB = pathB[0];
  let indexB = 0;

  // If A is deeper, skip indices that can't match.
  if (pathALength - pathBLength > 0) {
    indexA = pathALength - pathBLength;
    itemA = pathA[indexA];
    pathALength = pathBLength;
  }

  // If B is deeper, skip indices that can't match
  if (pathBLength - pathALength > 0) {
    indexB = pathBLength - pathALength;
    itemB = pathB[indexB];
    pathBLength = pathALength;
  }

  // Walk in lockstep until a match is found
  let depth = pathALength;
  while (depth--) {
    if (itemA === itemB) {
      return itemA;
    }
    itemA = pathA[indexA++];
    itemB = pathB[indexB++];
  }
  return null;
}

/**
 * Determine whether any of the active touches are within the current responder.
 * This cannot rely on W3C `targetTouches`, as Safari does not implement it.
 */
export function hasTargetTouches(target: any, touches: TouchList | undefined) {
  if (touches == null || touches.length === 0) {
    return false;
  }
  for (const touch of touches) {
    const node = touch.target;
    if (node != null) {
      if (target.contains(node)) {
        return true;
      }
    }
  }
  return false;
}

/**
 * Ignore `selectionchange` events that don't correspond with a person's intent
 * to select text.
 */
export function hasValidSelection(domEvent: Event) {
  if (domEvent.type === 'selectionchange') {
    return isSelectionValid();
  }
  return domEvent.type === 'select';
}

/**
 * Events are only valid if the primary button was used without specific
 * modifier keys.
 */
export function isPrimaryPointerDown(domEvent: any) {
  const touch = domEvent.type === 'touchstart' || domEvent.type === 'touchmove';
  const primaryMouseDown =
    domEvent.type === 'mousedown' && (domEvent.button === 0 || domEvent.buttons === 1);
  const primaryMouseMove = domEvent.type === 'mousemove' && domEvent.buttons === 1;
  const noModifiers = domEvent.altKey === false && domEvent.ctrlKey === false;

  if (touch || (primaryMouseDown && noModifiers) || (primaryMouseMove && noModifiers)) {
    return true;
  }
  return false;
}

export function isStartish(domEvent: Event): boolean {
  return domEvent.type === 'touchstart' || domEvent.type === 'mousedown';
}

export function isMoveish(domEvent: Event): boolean {
  return domEvent.type === 'touchmove' || domEvent.type === 'mousemove';
}

export function isEndish(domEvent: Event): boolean {
  return domEvent.type === 'touchend' || domEvent.type === 'mouseup' || isCancelish(domEvent);
}

export function isCancelish(domEvent: Event): boolean {
  return domEvent.type === 'touchcancel' || domEvent.type === 'dragstart';
}

export function isScroll(domEvent: Event): domEvent is Event {
  return domEvent.type === 'scroll';
}

export function isSelectionChange(domEvent: Event): domEvent is Event {
  return domEvent.type === 'select' || domEvent.type === 'selectionchange';
}
