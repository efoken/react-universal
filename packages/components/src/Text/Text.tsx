'use client';

import type { AnyProps } from '@react-universal/core';
import {
  forwardedProps,
  getLocaleDirection,
  styled,
  useElementLayout,
  useOwnerState,
  usePlatformMethods,
} from '@react-universal/core';
import { isString, pick } from '@react-universal/utils';
import { useComposedRefs } from '@tamagui/compose-refs';
import { forwardRef, useCallback, useContext, useRef } from 'react';
import type { GestureResponderEvent } from 'react-native';
import type { TextMethods, TextOwnerState, TextProps, TextType } from './Text.types';
import { TextAncestorContext } from './TextAncestorContext';

function pickProps<T extends AnyProps>(props: T) {
  return pick(props, {
    ...forwardedProps.defaultProps,
    ...forwardedProps.accessibilityProps,
    ...forwardedProps.clickProps,
    ...forwardedProps.focusProps,
    ...forwardedProps.keyboardProps,
    ...forwardedProps.mouseProps,
    ...forwardedProps.touchProps,
    ...forwardedProps.styleProps,
    href: true,
    lang: true,
  });
}

const TextRoot = styled('div', {
  name: 'Text',
  slot: 'Root',
})<{ ownerState: TextOwnerState }>(
  ({ ownerState, theme }) => ({
    backgroundColor: 'transparent',
    fontFamily: theme.fonts.body.family,
    textAlign: 'start' as any,
    whiteSpace: 'pre-wrap',
    wordWrap: 'break-word',
    ...(ownerState.hasTextAncestor && {
      fontFamily: 'inherit' as any,
      textAlign: 'inherit' as any,
      whiteSpace: 'inherit',
    }),
    ...(ownerState.pressable && {
      cursor: 'pointer',
    }),
  }),
  ({ ownerState }) =>
    ownerState.numberOfLines != null && {
      ...(ownerState.numberOfLines === 1 && {
        maxWidth: '100%',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap',
        wordWrap: 'normal',
      }),
      ...(ownerState.numberOfLines > 1 && {
        display: '-webkit-box' as any,
        maxWidth: '100%',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        WebkitBoxOrient: 'vertical',
        WebkitLineClamp: ownerState.numberOfLines,
      }),
    },
);

export const Text = forwardRef<HTMLElement & TextMethods, TextProps>(
  ({ as: _as, hrefAttrs, numberOfLines, onClick, onLayout, onPress, ...props }, ref) => {
    const hasTextAncestor = useContext(TextAncestorContext);
    const hostRef = useRef<HTMLElement>(null);

    useElementLayout(hostRef, onLayout);

    const handleClick = useCallback(
      (event: GestureResponderEvent) => {
        if (onClick != null) {
          onClick(event);
        } else if (onPress != null) {
          event.stopPropagation();
          onPress(event);
        }
      },
      [onClick, onPress],
    );

    let component: 'a' | 'div' | 'span' = hasTextAncestor ? 'span' : 'div';

    const langDirection = props.lang == null ? undefined : getLocaleDirection(props.lang);
    const componentDirection = props.dir ?? langDirection;

    const supportedProps: AnyProps = pickProps(props);
    supportedProps.dir = componentDirection;
    // 'auto' by default allows browsers to infer writing direction (root elements only)
    if (!hasTextAncestor) {
      supportedProps.dir = componentDirection ?? 'auto';
    }

    if (onClick != null || onPress != null) {
      supportedProps.onClick = handleClick;
    }

    if (props.href != null) {
      component = 'a';
      if (hrefAttrs != null) {
        const { download, rel, target } = hrefAttrs;
        if (download != null) {
          supportedProps.download = download;
        }
        if (rel != null) {
          supportedProps.rel = rel;
        }
        if (isString(target)) {
          supportedProps.target = target.startsWith('_') ? target : `_${target}`;
        }
      }
    }

    const platformMethodsRef = usePlatformMethods(hostRef);
    const handleRef = useComposedRefs<HTMLElement>(hostRef, platformMethodsRef, ref);

    supportedProps.ref = handleRef;

    const ownerState = useOwnerState({
      hasTextAncestor,
      numberOfLines,
      pressable: onClick != null || onPress != null,
      role: props.role,
    });

    const element = <TextRoot as={_as ?? component} ownerState={ownerState} {...supportedProps} />;

    return hasTextAncestor ? (
      element
    ) : (
      <TextAncestorContext.Provider value>{element}</TextAncestorContext.Provider>
    );
  },
) as TextType;

Text.displayName = 'Text';
