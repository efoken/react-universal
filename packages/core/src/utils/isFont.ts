import { isObject } from '@react-universal/utils';
import type { AnyObject } from '@react-universal/utils';
import type { ThemeFont } from '../theme';

export function isFont(obj: AnyObject): obj is ThemeFont {
  return isObject(obj) && 'family' in obj && 'weights' in obj;
}
