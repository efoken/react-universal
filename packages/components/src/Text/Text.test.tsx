import type { AnyObject } from '@react-universal/utils';
import { isFunction, noop } from '@react-universal/utils';
import { act, fireEvent, render } from '@testing-library/react';
import { createRef } from 'react';
import { describe, expect, test, vi } from 'vitest';
import { Text } from './Text';
import type { TextType } from './Text.types';

function createEventTarget(node: any) {
  return {
    node,
    blur: () => fireEvent.blur(node),
    click: () => fireEvent.click(node),
    focus: (payload?: AnyObject) => fireEvent.focus(node, payload),
    pointerDown: (payload?: AnyObject) => fireEvent.pointerDown(node, payload),
  };
}

describe('Text', () => {
  test('default', () => {
    const { container } = render(<Text />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('nested', () => {
    const { container } = render(
      <Text>
        <Text testID="child" />
      </Text>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('prop "aria-label"', () => {
    test('value is set', () => {
      const { container } = render(<Text aria-label="accessibility label" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "aria-labelledby"', () => {
    test('value is set', () => {
      const { container } = render(<Text aria-labelledby="123" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "aria-live"', () => {
    test('value is set', () => {
      const { container } = render(<Text aria-live="polite" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "role"', () => {
    test('value is set', () => {
      const { container } = render(<Text role="presentation" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value is "button"', () => {
      const { container } = render(<Text role="button" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value alters HTML element', () => {
      const { container } = render(<Text role="article" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "dir"', () => {
    test('value is "ltr"', () => {
      const { container } = render(<Text dir="ltr" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value is "rtl"', () => {
      const { container } = render(<Text dir="rtl" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "href"', () => {
    test('value is set', () => {
      const { container } = render(<Text href="https://example.com" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('href with role', () => {
      const { container } = render(<Text role="presentation" href="https://example.com" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "hrefAttrs"', () => {
    test('requires "href"', () => {
      const { container } = render(<Text hrefAttrs={{ download: 'filename.jpg' }} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value is set', () => {
      const hrefAttrs = {
        download: 'filename.jpg',
        rel: 'nofollow',
        target: '_blank',
      };
      const { container } = render(<Text href="https://example.com" hrefAttrs={hrefAttrs} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('target variant is set', () => {
      const hrefAttrs = {
        target: 'blank',
      };
      const { container } = render(<Text href="https://example.com" hrefAttrs={hrefAttrs} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('null values are excluded', () => {
      const hrefAttrs = {
        download: undefined,
        rel: undefined,
        target: undefined,
      };
      const { container } = render(<Text href="https://example.com" hrefAttrs={hrefAttrs} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "lang"', () => {
    test('undefined', () => {
      const { container } = render(<Text />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('fr', () => {
      const { container } = render(<Text lang="fr" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('ar', () => {
      const { container } = render(<Text lang="ar" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('with dir', () => {
      const { container } = render(<Text dir="ltr" lang="ar" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "id"', () => {
    test('value is set', () => {
      const { container } = render(<Text id="ID" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "numberOfLines"', () => {
    test('value is set', () => {
      const { container } = render(<Text numberOfLines={3} />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test('value is set to one', () => {
      const { container } = render(<Text numberOfLines={1} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "onBlur"', () => {
    test('is called', () => {
      const onBlur = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        // @ts-expect-error: `onBlur` is Web only and does not exist in types
        render(<Text ref={ref} onBlur={onBlur} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.focus();
        target.blur();
      });
      expect(onBlur).toBeCalled();
    });
  });

  describe('prop "onClick"', () => {
    test('is called', () => {
      const onClick = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        render(<Text ref={ref} onClick={onClick} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.click();
      });
      expect(onClick).toBeCalled();
    });

    test('is still called if "onPress" is provided', () => {
      const onClick = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        render(<Text ref={ref} onClick={onClick} onPress={noop} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.click();
      });
      expect(onClick).toBeCalled();
    });
  });

  describe('prop "onFocus"', () => {
    test('is called', () => {
      const onFocus = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        // @ts-expect-error: `onFocus` is Web only and does not exist in types
        render(<Text ref={ref} onFocus={onFocus} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.focus();
      });
      expect(onFocus).toBeCalled();
    });
  });

  describe('prop "onPointerDown"', () => {
    // beforeEach(() => {
    //   setPointerEvent(true);
    // });
    // afterEach(() => {
    //   setPointerEvent(false);
    // });

    test('is called', () => {
      const onPointerDown = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        // @ts-expect-error: `onPointerDown` is Web only and does not exist in types
        render(<Text ref={ref} onPointerDown={onPointerDown} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.pointerDown({ pointerType: 'touch' });
      });
      expect(onPointerDown).toBeCalled();
    });
  });

  describe('prop "onPress"', () => {
    test('is called', () => {
      const onPress = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        render(<Text ref={ref} onPress={onPress} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.click();
      });
      expect(onPress).toBeCalled();
    });

    test('is not called if "onClick" is provided', () => {
      const onPress = vi.fn();
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        render(<Text ref={ref} onClick={noop} onPress={onPress} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.click();
      });
      expect(onPress).not.toBeCalled();
    });
  });

  describe('prop "ref"', () => {
    test('value is set', () => {
      const ref = vi.fn();
      render(<Text ref={ref} />);
      expect(ref).toBeCalled();
    });

    test('is not called for prop changes', () => {
      const ref = vi.fn();
      let rerender: (ui: React.ReactNode) => void;
      act(() => {
        ({ rerender } = render(<Text ref={ref} id="123" style={{ borderWidth: 5 }} />));
      });
      expect(ref).toHaveBeenCalledTimes(1);
      act(() => {
        rerender(<Text ref={ref} id="1234" style={{ borderWidth: 6 }} />);
      });
      expect(ref).toHaveBeenCalledTimes(1);
    });

    test('node has imperative methods', () => {
      const ref = createRef<React.ElementRef<TextType>>();
      act(() => {
        render(<Text ref={ref} />);
      });
      const node = ref.current!;
      expect(isFunction(node.measure));
      expect(isFunction(node.measureLayout));
      expect(isFunction(node.measureInWindow));
    });
  });

  describe('prop "style"', () => {
    test('value is set', () => {
      const { container } = render(<Text style={{ borderWidth: 5 }} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "testID"', () => {
    test('value is set', () => {
      const { container } = render(<Text testID="123" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
