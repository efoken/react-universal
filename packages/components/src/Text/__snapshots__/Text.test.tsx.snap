// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text > default 1`] = `
<div
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > nested 1`] = `
<div
  class="css-8myfei"
  dir="auto"
>
  <span
    class="css-1ii5pxg"
    data-testid="child"
  />
</div>
`;

exports[`Text > prop "aria-label" > value is set 1`] = `
<div
  aria-label="accessibility label"
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > prop "aria-labelledby" > value is set 1`] = `
<div
  aria-labelledby="123"
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > prop "aria-live" > value is set 1`] = `
<div
  aria-live="polite"
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > prop "dir" > value is "ltr" 1`] = `
<div
  class="css-8myfei"
  dir="ltr"
/>
`;

exports[`Text > prop "dir" > value is "rtl" 1`] = `
<div
  class="css-8myfei"
  dir="rtl"
/>
`;

exports[`Text > prop "href" > href with role 1`] = `
<a
  class="css-8myfei"
  dir="auto"
  href="https://example.com"
  role="presentation"
/>
`;

exports[`Text > prop "href" > value is set 1`] = `
<a
  class="css-8myfei"
  dir="auto"
  href="https://example.com"
/>
`;

exports[`Text > prop "hrefAttrs" > null values are excluded 1`] = `
<a
  class="css-8myfei"
  dir="auto"
  href="https://example.com"
/>
`;

exports[`Text > prop "hrefAttrs" > requires "href" 1`] = `
<div
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > prop "hrefAttrs" > target variant is set 1`] = `
<a
  class="css-8myfei"
  dir="auto"
  href="https://example.com"
  target="_blank"
/>
`;

exports[`Text > prop "hrefAttrs" > value is set 1`] = `
<a
  class="css-8myfei"
  dir="auto"
  download="filename.jpg"
  href="https://example.com"
  rel="nofollow"
  target="_blank"
/>
`;

exports[`Text > prop "id" > value is set 1`] = `
<div
  class="css-8myfei"
  dir="auto"
  id="ID"
/>
`;

exports[`Text > prop "lang" > ar 1`] = `
<div
  class="css-8myfei"
  dir="rtl"
  lang="ar"
/>
`;

exports[`Text > prop "lang" > fr 1`] = `
<div
  class="css-8myfei"
  dir="ltr"
  lang="fr"
/>
`;

exports[`Text > prop "lang" > undefined 1`] = `
<div
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > prop "lang" > with dir 1`] = `
<div
  class="css-8myfei"
  dir="ltr"
  lang="ar"
/>
`;

exports[`Text > prop "numberOfLines" > value is set 1`] = `
<div
  class="css-18ravd1"
  dir="auto"
  style="--number-of-lines: 3;"
/>
`;

exports[`Text > prop "numberOfLines" > value is set to one 1`] = `
<div
  class="css-1xgxtno"
  dir="auto"
  style="--number-of-lines: 1;"
/>
`;

exports[`Text > prop "role" > value alters HTML element 1`] = `
<article
  class="css-8myfei"
  dir="auto"
/>
`;

exports[`Text > prop "role" > value is "button" 1`] = `
<button
  class="css-8myfei"
  dir="auto"
  type="button"
/>
`;

exports[`Text > prop "role" > value is set 1`] = `
<div
  class="css-8myfei"
  dir="auto"
  role="presentation"
/>
`;

exports[`Text > prop "style" > value is set 1`] = `
<div
  class="css-8myfei"
  dir="auto"
  style="border-width: 5px;"
/>
`;

exports[`Text > prop "testID" > value is set 1`] = `
<div
  class="css-8myfei"
  data-testid="123"
  dir="auto"
/>
`;
